[US2814] Implementation UNMAP functionality in zfs

mynktl opened the pull request:
Changes: 
- Handling of UNMAP command in ZREPL.
- Changes to support the rebuilding of UNMAP region in ZREPL.
- Modification in `uzfs_read_data` to handle read of blocks having no data(HOLES).

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
[TA2921] Implementation UNMAP functionality in uzfs core level
[TA2922] Implementation of test utility to verify read/write/unmap region

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
- Handling of UNMAP command from target
- Avoiding error for uzfs_read_data in test code

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Merge remote-tracking branch 'remote/zfs-0.7-release' into unmap

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
[TA2936] [TA2923] Implementation of rebuilding unmap region and unit test.

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

mynktl committed:
Merge branch 'zfs-0.7-release' into unmap

Commit:
Travis build fix

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Merge branch 'unmap' of github.com:mynktl/zfs into unmap

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -512,3 +536,122 @@ uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t tgt_block_size)
+			count--;
+	}
+
+	error = dmu_free_long_range(zv->zv_objset, ZVOL_OBJ, offset, len);
```
its better to unmap and write IO num in same tx

mynktl replied:
Ok.

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -512,3 +536,122 @@ uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t tgt_block_size)
+	}
+
+	if (!zil_replaying(zv->zv_zilog, tx)) {
+		itx = zil_itx_create(TX_TRUNCATE, sizeof (*lr));
```
where are we storing io_num in the zil?

mynktl replied:
Previously, we were not storing io_num in zil. In the new implementation, we are storing io_num in zil record.

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -512,3 +536,122 @@ uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t tgt_block_size)
 }
+
+int
+uzfs_unmap_data(zvol_state_t *zv, uint64_t offset, uint64_t len,
```
this function can be merged with uzfs_write_data..?

mynktl replied:
yes. Now, we can.

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -377,6 +388,18 @@ uzfs_zvol_worker(void *arg)
 	}
+
+	if (rebuild_cmd_req && (hdr->opcode == ZVOL_OPCODE_READ) &&
+	    (data_err == EIO) && (!md_err)) {
```
EIO? It should be ENOENT right..?

mynktl replied:
Yes.
It should be `ENOENT`.
`dmu_buf_hold_array_by_dnode` returns `EIO` if `dbuf_hold_impl` returns db=NULL (dh->dh_err will be set to ENOENT in `dbuf_hold_impl`).
This is check is removed as per new implementation.

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -307,20 +319,32 @@ uzfs_read_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
-		error = dmu_read(os, ZVOL_OBJ, offset, bytes, buf + read, 0);
-		if (error != 0)
-			goto exit;
+		error = dmu_read(os, ZVOL_OBJ, offset, bytes,
```
what about the data in the buffer, in case of holes? and, how does caller knows which blocks are holes in case of data_err? what if entire block is not hole, but, only a part of it?

mynktl replied:
If part of block gets unmapped then it will not be declared as a hole. As per new implementation, This change will be removed.

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -294,6 +302,10 @@ uzfs_read_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 		len_in_first_aligned_block = len;
 
+#ifdef	_UZFS
+	dmu_read_flag = DMU_READ_CHECK_HOLE;
```
why do we need to set this flag all the time? setting this flag should be good enough if the read is for rebuild purpose

mynktl replied:
Ok. As per new implementation, This change will be removed.

Commit:
[TA2922] [TA2921] [TA2936] [TA2923] Implementation of UNMAP command in ZREPL with io_sequence number logic

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on module/zfs/dmu.c:
```diff
@@ -492,7 +496,17 @@ dmu_buf_hold_array_by_dnode(dnode_t *dn, uint64_t offset, uint64_t length,
 	for (i = 0; i < nblks; i++) {
-		dmu_buf_impl_t *db = dbuf_hold(dn, blkid + i, tag);
+		dmu_buf_impl_t *db = NULL;
+#ifdef	_UZFS
```
@mynktl .. are we not removing this old code?

mynktl replied:
This is not needed now. I will remove hole related thing added for unmap.

Commit:
Merge remote-tracking branch 'remote/zfs-0.7-release' into unmap

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on include/zrepl_mgmt.h:
```diff
@@ -258,6 +260,19 @@ uzfs_zinfo_take_refcnt(zvol_info_t *zinfo)
+/*
+ * Get unmap io sequence number
+ */
+#define	DECODE_IF_UNMAP_IOSEQ(_x)	((_x) &  (~(1ULL <<  63)))
```
lets make it as GET_IOSEQ

mynktl replied:
ok

vishnuitta commented on include/zrepl_mgmt.h:
```diff
@@ -258,6 +260,19 @@ uzfs_zinfo_take_refcnt(zvol_info_t *zinfo)
+/*
+ * Check if io sequence number is meant for unmap
+ */
+#define	CHECK_UNMAP_IOSEQ(_x)		((_x >> 63) & 1)
```
IS_UNMAP_IOSEQ

mynktl replied:
ok

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -361,6 +362,19 @@ uzfs_zvol_worker(void *arg)
 			break;
 
+		case ZVOL_OPCODE_UNMAP:
+			if ((rebuild_cmd_req &&
```
why only for rebuild_cmd_req?

mynktl replied:
it is for both.. for rebuild and also for app IOs

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -513,3 +513,81 @@ uzfs_update_metadata_granularity(zvol_state_t *zv, uint64_t tgt_block_size)
+chunk_io:
+			GET_NEXT_CHUNK(chunk_io, offset, len, end);
+			wrote = offset - orig_offset;
+			CHECK_FIRST_ALIGNED_BLOCK(
```
alignment here is not required, as, dnode_free_range takes care of it

mynktl replied:
ok

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -379,9 +393,11 @@ uzfs_zvol_worker(void *arg)
-	if (rebuild_cmd_req && (hdr->opcode == ZVOL_OPCODE_WRITE)) {
+	if (rebuild_cmd_req && ((hdr->opcode == ZVOL_OPCODE_WRITE) ||
+	    ((hdr->opcode == ZVOL_OPCODE_UNMAP) &&
+	    ZVOL_IS_REBUILDING(zinfo->main_zv)))) {
```
this check is not required right?

mynktl replied:
yes.. it can be removed since `rebuild_cmd_req` check is also there.

Commit:
- Added uzfs_perform_ops function to handle uzfs_write_data/uzfs_unmap_data
- Added uzfs_write_metadata function to write metadata to ZVOL_META_OBJ
- Travis build fix
- Added new opcodes for rebuild process
	ZVOL_OPCODE_H_RB_READ
	ZVOL_OPCODE_H_RB_UNMAP
	ZVOL_OPCODE_RB_WRITE
	ZVOL_OPCODE_RB_UNMAP
- Changed CHECK_UNMAP_IOSEQ/DECODE_IF_UNMAP_IOSEQ to IS_UNMAP_IOSEQ/GET_IOSEQ

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

mynktl committed:
Merge branch 'zfs-0.7-release' into unmap

Commit:
- Travis build fix
- Added env variable(avoid_rebuild_count_check) to avoid rebuild_op_cnt related
  check for START_REBUILD opcode

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

satbirchhikara commented on lib/libzpool/uzfs_io.c:
```diff
@@ -183,7 +116,89 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
+	    len == 0)
+		return (EINVAL);
+
+	if (offset + len > zv->zv_volsize || offset > zv->zv_volsize)
```
If offset is greater than volsize then (offset+ len) would also be greater than volsize. I think second part of this check is not required.

mynktl replied:
yes. It can be removed.

Commit:
- Unit test for replay of TX_TRUNCATE record
- fixing error in testing code

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

mynktl committed:
Merge branch 'zfs-0.7-release' into unmap

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -333,7 +397,7 @@ uzfs_zvol_worker(void *arg)
 		hdr->status = ZVOL_OP_STATUS_FAILED;
 		hdr->len = 0;
-		if (!(rebuild_cmd_req && (hdr->opcode == ZVOL_OPCODE_WRITE)))
+		if (!IS_IO_ACK_NEEDED(hdr))
```
here, `!` need to be removed

mynktl replied:
yes

vishnuitta commented on lib/libzrepl/data_conn.c:
```diff
@@ -320,9 +383,10 @@ uzfs_zvol_worker(void *arg)
 			zio_cmd_free(&zio_cmd);
-		if (hdr->opcode == ZVOL_OPCODE_WRITE)
+		if (hdr->opcode == ZVOL_OPCODE_WRITE ||
+		    hdr->opcode == ZVOL_OPCODE_RB_WRITE)
```
lets make new opcodes as:
        ZVOL_OPCODE_HELPING_READ,
        ZVOL_OPCODE_HELPING_UNMAP,
        ZVOL_OPCODE_DW_WRITE,
        ZVOL_OPCODE_DW_UNMAP

mynktl replied:
ok

vishnuitta commented on module/zfs/zvol.c:
```diff
@@ -622,7 +622,12 @@ zvol_replay_truncate(zvol_state_t *zv, lr_truncate_t *lr, boolean_t byteswap)
+#if defined(_KERNEL)
 	return (dmu_free_long_range(zv->zv_objset, ZVOL_OBJ, offset, length));
+#else
+	return (dmu_free_long_range_impl(zv->zv_objset, zv->zv_dn, offset,
```
here, version of lr can be checked, and pass NULL, NULL if version is VERSION_0

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -167,12 +100,12 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 
 		if (metadata)
-			WRITE_METADATA(zv, metablk, mdata, tx);
+			uzfs_write_metadata(zv, offset, bytes, metadata, tx);
```
lets handle the error here

mynktl replied:
ok

vishnuitta commented on module/zfs/dmu.c:
```diff
@@ -736,22 +741,55 @@ dmu_objset_zfs_unmounting(objset_t *os)
+				dmu_tx_abort(tx);
+				return (err);
+			}
+			err = uzfs_write_metadata(zv, offset, length,
```
lets error out by doing dmu_tx_abort her

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -44,9 +44,11 @@
 		return (-1);
-	if (first->io_num == second->io_num)
+	if (GET_IOSEQ(first->io_num) ==
+	    GET_IOSEQ(second->io_num))
```
do GET_IOSEQ in iszero() below function

mynktl replied:
ok

vishnuitta reviewed (commented):
as a general practice, I'm thinking to have required metadata of IO numbers with bit set within zv layer. To zinfo, lets have bit unset whenever required.
So, uzfs_read_data should unset the metadata. But, uzfs_read_metadata would need with bit set for the purpose of rebuild.
checkpointed io_num (degraded, healthy and running) are stored without bit set.

Commit:
- Addressing review comments
- Travis build fix

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Merge branch 'unmap' of github.com:mynktl/zfs into unmap

mynktl committed:
Merge branch 'zfs-0.7-release' into unmap

Commit:
- Changes to use unmap io sequence number at core(zv related) API only
- Some missed changes in testing code

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Merge branch 'unmap' of github.com:mynktl/zfs into unmap

Commit:
Cstyle fix

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Fix for race condition between new data connection and old connection cleanup

Signed-off-by: mayank <mayank.patel@cloudbyte.com>
